LauraThompson/Lecture3Exercise.py

```diff
@@ -0,0 +1,107 @@
+
+def find_longest_word(fruitlist):
+    longest_word = ' '
+    for word in fruitlist:
+        if len(word) > len(longest_word):
+            longest_word = word
+    print(longest_word)
+
+find_longest_word(fruitlist)
+
+#%%Question 2: Write a Python program to count the number of unique characters in a string
+myname = "My name is Laura Thompson"
+#now count it.
+#first counting with both upper and lower case a unique
+allcount=len(set(myname))
+print("The number of characters with upper and lower unique is:", allcount)
+#set to all lower case so that letters of upper and lower aren't unique.
+lowercount=len(set(myname.lower()))
+print("The number of characters without case sensitivity is:", lowercount)
+
+#%%Question 3: Wrtie a Python program to sort (ascending and descending) a dictionary by value
+
+dict1 = {"first": "Laura", "middleinitial": "J", "last": "Thompson"}
+#print(dict1)
+#print(dict1.keys())
+#print(dict1.values())
+dict1["birthstate"]="Nebraska"
+#print(dict1)
+
+#now sort it.
+sorted_ascending_dict1 = sorted(dict1)
+print(sorted_ascending_dict1)
+
+sorted_descending_dict1 = sorted(dict1, reverse=True)
+print(sorted_descending_dict1)
+
+#%%Question 4: Write a Python program to change a given string to a new string where the first and last chars have been exchanged
+dogstory = "The little dog jumped over the large, tall fence"
+
+#now switch it.
+dogstory[1:-1]
+reversedogstory=dogstory[-1:]+dogstory[1:-1]+dogstory[:1]
+print(reversedogstory)
+#%%Question 5: Write a Python function that takes two lists and returns True if they havea least one common member
+
+cropslist = ["corn", "soybean", "wheat", "sorghum"]
+grocerylist = ["milk", "apples", "flour", "corn"]
+
+def common_elements(cropslist, grocerylist):
+    result = []
+    for element in cropslist:
+        if element in grocerylist:
+            result.append(element)
+    return result
+
+print(common_elements(cropslist, grocerylist))
+
+#%%Question 6: Write a Python function that returns the minimum number of coins ($0.01, $0.1, $0.25, $1) that make a given value.
+#from math import remainder
+#from decimal import decimal
+# more complicated...first needs to divide by 1, then 0.25, then 0.1, then 0.01... or something like that...
+#print(Decimal('3.5')%Decimal('0.1'))
+#print(4.5%1)
+#print(remainder(4.5,1))
+#CurrencyValue=7.75
+#CurrencyValue/1
+#rem=rem%1;
+
+#greedy method example with options of $.25, $.10, $.05, and $0.01.
+#def num_coins(cents):
+ #   coins = [25,10,5,1]
+  #  count = 0
+   # for coin in coins:
+    #    while cents >= coin:
+     #       cents = cents - coin
+      #      count = count + 1
+            
+#    return count
+#print(num_coins(10))
+
+
+#greedy method
+#but we did not have $0.05 as an option, and also have whole dollar options. So adding 100, and removing 5.
+def num_coins(cents):
+    cents = cents*100
+    coins = [100,25,10,1]
+    count = 0
+    for coin in coins:
+        while cents >= coin:
```

Comment:
You can use a floor division to calculate the number of `coin` to avoid the while-loop. 
```
count = cents // coin
cents -+ count * coin
```

Reply:
Thank you for the suggestion. I was not aware of these three operators.

Can these three lines of code be used in a function? So that someone can just put in any value and get the result?

When I try this code I get the following result: TypeError: unsupported operand type(s) for //: 'float' and 'list'

Googled the three operands used here. Notes below for my own reference.
// floor division rounds off the decimals
-+ addition and subtraction?
+= assigns the value on the right to the variable on the left.

Reply:
`-+` should be `-=` we talked about these operators in Lecture 3.

Reply:
I tried this suggestion as follows:

```
#alternate method without while loop
def num_coins(cents):
    cents = cents*100
    coin = [100,25,10,1]
    count = 0
    coin_count = cents // coin
    cents -= coin_count * coin
    count += coin_count
print(num_coins(5.27))
```

I get an error:
TypeError: unsupported operand type(s) for //: 'float' and 'list'


Note to self:
-= subtract right operand from the left operand and assign the result to the left operand.
+=adds the right operand to the left operand and assigns the result to left operand.

Reply:
You still need to iterate the coins
```python
coins = [100,25,10,1]
count = 0
for coin in coins:
    coin_count = cents // coin
    cents -= coin_count * coin
    count += coin_count
```